src/__tests__/types.typescript.tsx

```diff
@@ -0,0 +1,308 @@
+  actions,
+  name: 'Type',
+});
+
+/**
+ * createSubscriber types tests
+ */
+
+const TypeSubscriber = createSubscriber<State, Actions>(TypeStore);
+
+Test = (
+  // @ts-ignore Child arg shape should be state + actions
+  <TypeSubscriber>{({ foo }) => foo}</TypeSubscriber>
+);
+
+Test = (
+  // @ts-ignore Actions should be correcly typed
+  <TypeSubscriber>{(__, { increment }) => increment()}</TypeSubscriber>
+);
+
+// Correct
+Test = <TypeSubscriber>{({ count }) => count + 0}</TypeSubscriber>;
+Test = <TypeSubscriber>{(__, { increment }) => increment(1)}</TypeSubscriber>;
+
+/**
+ * createSubscriber with selector types tests
+ */
+const TypeSelector = createSubscriber<State, Actions, { baz: number }>(
+  TypeStore,
+  {
+    selector: state => ({ baz: 1 }),
+  }
+);
+
+Test = (
+  // @ts-ignore Child arg shape should be pick + actions
+  <TypeSelector>{({ count }) => count}</TypeSelector>
+);
+
+Test = (
+  // @ts-ignore Should not accept props
+  <TypeSelector min={3}>{({ baz }) => baz}</TypeSelector>
+);
+
+// Correct
+Test = <TypeSelector>{({ baz }) => baz}</TypeSelector>;
+Test = <TypeSelector>{(__, { increment }) => increment(1)}</TypeSelector>;
+
+const TypeSelectorNull = createSubscriber<State, Actions, void>(TypeStore, {
+  selector: null,
+});
+
+Test = (
+  // @ts-ignore Child arg shape should be just actions
+  <TypeSelectorNull>{({ count }) => count}</TypeSelectorNull>
+);
+
+// @ts-ignore Should not accept props
+Test = <TypeSelectorNull myProp>{state => !state}</TypeSelectorNull>;
+
+// Correct
+Test = (
+  <TypeSelectorNull>{(state, { increment }) => increment(1)}</TypeSelectorNull>
+);
+
+type SelectorState = { baz: number; min: number };
+
+const TypeSelectorProp = createSubscriber<
+  State,
+  Actions,
+  SelectorState,
+  SelectorProps
+>(TypeStore, {
+  selector: (state, props) => ({ baz: 1, min: props.min }),
+});
+
+Test = (
+  // @ts-ignore Should require props
+  <TypeSelectorProp>{({ baz }) => baz}</TypeSelectorProp>
+);
+
+Test = (
+  // @ts-ignore Should require correct prop types
+  <TypeSelectorProp min="2">{({ baz }) => baz}</TypeSelectorProp>
+);
+
+Test = (
+  // @ts-ignore Should have correct selector types
+  <TypeSelectorProp min={2}>{({ min }) => min.split('')}</TypeSelectorProp>
+);
+
+// Correct
+Test = (
+  <TypeSelectorProp min={2}>{({ baz, min }) => baz + min}</TypeSelectorProp>
+);
+
+/**
+ * createHook types tests
+ */
+
+const typeBaseHook = createHook<State, Actions>(TypeStore);
+
+const baseReturn = typeBaseHook();
+
+// @ts-ignore Should not accept arguments
+typeBaseHook({});
+
+// @ts-ignore Array index 0 should be state
+baseReturn[0].foo;
+
+// @ts-ignore Array index 1 should be actions
+baseReturn[1].increment();
+
+// @ts-ignore Array index 1 should be actions
+baseReturn[1].increment('1');
+
+// @ts-ignor
+baseReturn[1].decrement().then(v => v);
+
+// Correct
+baseReturn[0].count + 0;
+baseReturn[1].increment(1);
+baseReturn[1].decrement();
+baseReturn[1].fetch().then(v => v);
```

Comment:
There is something off with the types here. TS says that the return of all those functions is `any` instead of the actual return type. And indeed in `.then(v => v)`, `v` is type `any` instead of `string`

Reply:
it seems to be due to StateAction "redefinition"

Reply:
Fix this. I still see an issue when directly specifying actions as a literal in createStore.